doc/design/draft/subscription.md

```diff
@@ -0,0 +1,180 @@
+| P1        | 31        | D      |
+| P2        | 24        | H      |
+| P2        | 22        | G      |
+| P1        | 21        | C      |
+| P2        | 20        | F      |
+| P1        | 11        | B      |
+| P1        | 1         | A      |
+
+### Subscribing with concurrent send
+
+Let's assume publisher P1 would like to deliver the sample K at timepoint
+52 while the subscriber requested subsription. Furthermore, the subscriber
+requests always a _HistorySize_ of 2.
+
+All the listed cases should behave identical if there is more then one
+concurrent publish in progress.
+
+#### Publisher before _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+
+The sample K will be delivered after successful subscription.
+
+#### Publisher after _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P1        | 51        | K      |
+| P2        | 42        | J      |
+
+The sample K will not be delivered by the publisher since the _DeliveryTime_
+is defined as exactly the timepoint after the sample is placed into the ChunkHistory
+and the subscriber cache is based on that ChunkHistory.
+
+### Subscribing After Publisher Stops Offering
+
+In this case we assume that publisher P1 has stopped offering the service
+before the subscription was requested.
+
+#### Subscription with _HistorySize_ 4
+
+The subscriber cache contains the four samples ordered after _DeliveryTime_.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+
+
+
+## Design
+
+### Idea 1: One Ring Buffer Which Contains The Samples
+
+- All publishers are sharing one random access lock-free ring-buffer.
+- The ring-buffer is filled with shared-chunks which can be copied in a threadsafe
+    and lock-free manner.
+- A publisher delivers a sample by writing that sample into the shared ring-buffer.
+- Every subscriber owns an index to their own read position of the ring-buffer.
+- On subscription the subscriber will acquire a const reference to the ring-buffer.
+- When subscriber acquires a sample with take, the shared-chunk is copied from
+    the index position and the index is incremented.
+- When the publisher delivers a sample it will verify all subscriber indices.
+   If a subscriber `index > ring-buffer-head - subscriberHistorySize` the index
+   is incremented.
+
+#### Open Points
+
+- Many chunks will not be released even when no susbcriber is using them.
+  - This might be solvable by having the ring-buffer size as the maximum of
+    history size and subscriber queue capacity?!
+- ring-buffer size? Since the ring-buffer is shared between multiple publishers
```

Comment:
Difficult. Theoretically it must have the size of the maximum of any currently requested history or offering publisher. This is not known in a totally dynamic world and will in addition require the buffer to always be at some maximum size (wasting shared memory if chunks are cached) or require it to dynamically grow (algorithmically difficult in a lock-free setting). 
If it is defined by the service options it would be better since then it is somewhat agreed on by all parties offering the service? Hard to say what is reasonable, usable and implementable.